src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/ProjectExtensions.cs

```diff
@@ -58,7 +59,110 @@ public static async ValueTask<bool> AppliesToProjectAsync(this Type type, IProje
+                return false;
+            }
+
+            // if the attribute is not applied then this type defaults to opt into language support
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type is a codefixer that should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature decorated with <see cref="ExportCodeFixProviderAttribute"/>.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the codefixer should be hidden from the user. Will return false for non-codefixer objects.</returns>
+        private static bool DoesCodeFixerFilterThisFeature(Type type, IProject project)
+        {
+            var analyzerAttr = type.CustomAttributes.FirstOrDefault(a => a.AttributeType.FullName.Equals(typeof(ExportCodeFixProviderAttribute).FullName, StringComparison.Ordinal));
+            if (analyzerAttr is not null)
+            {
+                var applicableLanguage = analyzerAttr.ConstructorArguments.First().Value as string;
+                if (applicableLanguage is not null && !string.IsNullOrWhiteSpace(applicableLanguage))
+                {
+                    if (project.Language == Language.CSharp
+                        && !applicableLanguage.Equals(CodeAnalysis.LanguageNames.CSharp, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+
+                    if (project.Language == Language.VisualBasic
+                        && !applicableLanguage.Equals(CodeAnalysis.LanguageNames.VisualBasic, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+
+                    if (project.Language == Language.FSharp
+                        && !applicableLanguage.Equals(CodeAnalysis.LanguageNames.FSharp, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Checks to see if the type has an ApplicableLanguageAttribute that should filter the feature out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature decorated with <see cref="ApplicableLanguageAttribute"/>.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the object should be hidden from the user. Will return false for objects not defning the language they support.</returns>
+        private static bool DoesApplicableLanguageFilterThisFeature(Type type, IProject project)
+        {
+            var applicableLangAttr = type.CustomAttributes.FirstOrDefault(a => a.AttributeType.FullName.Equals(typeof(ApplicableLanguageAttribute).FullName, StringComparison.Ordinal));
```

Comment:
We might want to allow multiple ApplicableLanguage attributes, but I'm not sure how much of a priority that is.

Reply:
At the moment, if something applies to multiple languages then it wouldn't use the ApplicableLanguage filter. The default is opt-in to all languages.

Decided that if we need to filter something that applies to VB and C# but not F# that would be a design decision, and unit tests, for later. LMK if you think it's worth considering as part of this change.